### PR TITLE
Hw spring boot security

### PR DESCRIPTION
**Pull request by SviatoslavTataryn**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -74,6 +74,11 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-validation</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>3.3.0</version>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@Configuration
+@AllArgsConstructor
+@EnableWebSecurity
+public class SecurityConfig {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/config/SecurityConfig.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@Configuration
+@AllArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/config/SecurityConfig.java]: Read about difference between annotations 
```suggestion
@RequiredArgsConstructor
```

File: src/main/java/mate/academy/bookstore/config/SecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+                .authorizeHttpRequests(authorizeRequests ->
+                        authorizeRequests
+                                .requestMatchers("/swagger-ui/**",
+                                        "/v3/api-docs/**",
+                                        "/swagger-resources/**",
+                                        "/webjars/**",
+                                        "/v2/api-docs/**") // дозволяє доступ до Swagger та OpenAPI
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/config/SecurityConfig.java]: remove comment

File: src/main/java/mate/academy/bookstore/controller/AuthenticationController.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import mate.academy.bookstore.service.UserService;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Authentication management",
+        description = "Endpoints for managing users")
+@AllArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthenticationController.java]:
```suggestion
@RequiredArgsConstructor
```

File: src/main/java/mate/academy/bookstore/controller/AuthenticationController.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+@RestController
+@RequestMapping("/auth")
+public class AuthenticationController {
+    private final UserService userService;
+
+    @Operation(summary = "Registration a new user",
+            description = "Endpoint to create a new user in the system.")
+    @PreAuthorize("hasAnyRole('USER')")
+    @PostMapping("/registration")
+    public UserResponseDto registerUser(@RequestBody UserRegistrationRequestDto requestDto
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthenticationController.java]:
```suggestion
    public UserResponseDto registerUser(@RequestBody @Valid UserRegistrationRequestDto requestDto
```

File: src/main/java/mate/academy/bookstore/controller/AuthenticationController.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        description = "Endpoints for managing users")
+@AllArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthenticationController {
+    private final UserService userService;
+
+    @Operation(summary = "Registration a new user",
+            description = "Endpoint to create a new user in the system.")
+    @PreAuthorize("hasAnyRole('USER')")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthenticationController.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        @NotBlank
+        @Email
+        String email,
+        @NotBlank
+        @Length(min = 8, max = 35)
+        String password,
+        @NotBlank
+        @Length(min = 8, max = 35)
+        String repeatPassword,
+        @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/user/UserRegistrationRequestDto.java]:
```suggestion
        @NotBlank
```

fot String type

File: src/main/java/mate/academy/bookstore/exception/UserNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.bookstore.exception;
+
+public class UserNotFoundException extends RuntimeException {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/exception/UserNotFoundException.java]: remove

File: src/main/java/mate/academy/bookstore/mapper/UserMapper.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import mate.academy.bookstore.dto.user.UserResponseDto;
+import mate.academy.bookstore.model.Role;
+import mate.academy.bookstore.model.User;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.Named;
+
+@Mapper(config = MapperConfig.class)
+public interface UserMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/mapper/UserMapper.java:12]: You need only two methods

```
User toModel(UserRegistrationRequestDto user);

UserResponseDto toUserResponse(User user);
```

File: src/main/java/mate/academy/bookstore/model/Role.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    @Enumerated(EnumType.STRING)
+    @Column(name = "name", nullable = false, unique = true)
+    private RoleName name;
+
+    public Role() {
+    }
+
+    public Role(RoleName name) {
+        this.name = name;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Role.java]:
```suggestion
    @Override
    public String getAuthority() {
        return name.name();
    }
```

File: src/main/java/mate/academy/bookstore/model/Role.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+
+@Entity
+@Data
+@Table(name = "roles")
+public class Role {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Role.java]:
```suggestion
public class Role implements GrantedAuthority {
```

File: src/main/java/mate/academy/bookstore/model/Role.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "name", nullable = false, unique = true)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Role.java]:
```suggestion
    @Column(nullable = false, unique = true)
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    private String password;
+    @Column(nullable = false)
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+    @Column(nullable = false)
+    private String shippingAddress;
+    @Column(name = "is_deleted")
+    private boolean isDeleted;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    @JoinTable(name = "user_roles",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @JoinTable(name = "users_roles",
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            inverseJoinColumns = @JoinColumn(name = "role_id"))
+    private Set<Role> roles = new HashSet<>();
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        List<GrantedAuthority> authorityList = new ArrayList<>();
+        for (Role role : roles) {
+            authorityList.add(new SimpleGrantedAuthority(role.getName().name()));
+        }
+        return authorityList;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
        return roles;
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    public void addRole(Role role) {
+        roles.add(role);
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/security/UserDetailsService.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.RequiredArgsConstructor;
+import mate.academy.bookstore.repository.UserRepository;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class UserDetailsService
+        implements org.springframework.security.core.userdetails.UserDetailsService {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/UserDetailsService.java]:
```suggestion
public class UserDetailsService implements UserDetailsService {
```

File: src/main/java/mate/academy/bookstore/security/UserDetailsService.java [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email).orElseThrow(() ->
+                new UsernameNotFoundException("Can't find user by email" + email));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/UserDetailsService.java]:
```suggestion
                new EntityNotFoundException("Can't find user by email" + email));
```

File: src/main/java/mate/academy/bookstore/security/UserDetailsService.java [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class UserDetailsService
+        implements org.springframework.security.core.userdetails.UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/UserDetailsService.java]:
```suggestion
    public UserDetails loadUserByUsername(String email) {
```

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+@AllArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        if (userRepository.findByEmail(requestDto.email()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(requestDto.email())) {
```

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+        Role role = roleRepository.findByName(
+                Role.RoleName.valueOf(requestDto.role().toUpperCase()));
+        if (role == null) {
+            throw new RuntimeException("Role not found");
+        }
+        user.addRole(role);
+
+        User savedUser = userRepository.save(user);
+        return userMapper.toUserResponse(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java]:
```suggestion
        user.setPassword(passwordEncoder.encode(requestDto.password()));

        Role role = roleRepository.findByName(Role.RoleName.USER).orElseThrow(......;
        user.setRoles(Set.of(role));
        
        userRepository.save(user);
        return userMapper.toUserResponse(user);
```

File: src/main/resources/db/changelog/changes/04-create-user-roles-table.yaml [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+      changes:
+        - preConditions:
+            onFail: CONTINUE
+            onError: CONTINUE
+            preConditions:
+              - not:
+                  tableExists:
+                    tableName: user_roles
+        - createTable:
+            tableName: user_roles
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-user-roles-table.yaml]:
```suggestion
            tableName: users_roles
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job. Let's improve it

File: src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+import mate.academy.bookstore.repository.UserRepository;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java:14]:
```suggestion
```

**Review by MentorMateAcademy — APPROVED** [review]

**Pull request merged**